src/Symfony/Component/Translation/LocaleSwitcher.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation;
+
+use Symfony\Component\Routing\RequestContext;
+use Symfony\Contracts\Translation\LocaleAwareInterface;
+
+/**
+ * @author Kevin Bond <kevinbond@gmail.com>
+ */
+final class LocaleSwitcher implements LocaleAwareInterface
+{
+    /**
+     * @param LocaleAwareInterface[] $localeAwareServices
+     */
+    public function __construct(
+        private string $locale,
+        private iterable $localeAwareServices,
+        private ?RequestContext $requestContext = null
+    ) {
+    }
+
+    public function setLocale(string $locale): void
+    {
+        \Locale::setDefault($locale);
+
+        foreach ($this->localeAwareServices as $service) {
+            $service->setLocale($locale);
+        }
+
+        if ($this->requestContext) {
+            $this->requestContext->setParameter('_locale', $locale);
+        }
+    }
+
+    public function getLocale(): string
+    {
+        return $this->locale;
+    }
+
+    /**
+     * Switch to a new locale, execute a callback, then switch back to the original.
+     *
+     * @param callable():void $callback
+     */
+    public function switchLocale(string $locale, callable $callback): void
```

Comment:
What about `withLocale()` as this method's name?

Reply:
it can also imply a "wither" method :thinking: 

`inLocale(), scopeLocale()` maybe

Reply:
To me, `switchLocale` means that the switch is permanent. I'd go for `withLocale()`. If this is considered confusing because this is not a "wither method", maybe go for `runWithLocale($locale, $callback)` ?